test/Npgsql.Tests/Support/PgPostmasterMock.cs

```diff
@@ -42,42 +51,64 @@ internal PgPostmasterMock(string? connectionString = null)
-                var readBuffer = new NpgsqlReadBuffer(null!, stream, clientSocket, ReadBufferSize, Encoding, RelaxedEncoding);
-                var writeBuffer = new NpgsqlWriteBuffer(null!, stream, clientSocket, WriteBufferSize, Encoding);
+                // ReSharper disable once FunctionNeverReturns
+            }
+        }
 
-                await readBuffer.EnsureAsync(4);
-                var len = readBuffer.ReadInt32();
-                await readBuffer.EnsureAsync(len - 4);
+        internal async Task<ServerOrCancellationRequest> Accept()
+        {
+            var clientSocket = await _socket.AcceptAsync();
 
-                if (readBuffer.ReadInt32() == CancelRequestCode)
-                {
-                    _pendingCancellationRequests.Enqueue((readBuffer.ReadInt32(), readBuffer.ReadInt32()));
-                    readBuffer.Dispose();
-                    writeBuffer.Dispose();
-                    stream.Dispose();
-                    continue;
-                }
+            var stream = new NetworkStream(clientSocket, true);
+            var readBuffer = new NpgsqlReadBuffer(null!, stream, clientSocket, ReadBufferSize, Encoding,
+                RelaxedEncoding);
+            var writeBuffer = new NpgsqlWriteBuffer(null!, stream, clientSocket, WriteBufferSize, Encoding);
 
-                // This is not a cancellation, "spawn" a new server
-                readBuffer.ReadPosition -= 8;
-                var serverMock = new PgServerMock(stream, readBuffer, writeBuffer, ++_processIdCounter);
-                await serverMock.Startup();
+            await readBuffer.EnsureAsync(4);
+            var len = readBuffer.ReadInt32();
+            await readBuffer.EnsureAsync(len - 4);
 
-                _allServers.Add(serverMock);
-                _pendingServers.Enqueue(serverMock);
+            if (readBuffer.ReadInt32() == CancelRequestCode)
+            {
+                readBuffer.Dispose();
+                writeBuffer.Dispose();
+                stream.Dispose();
+                return new ServerOrCancellationRequest((readBuffer.ReadInt32(), readBuffer.ReadInt32()));
             }
 
-            // ReSharper disable once FunctionNeverReturns
+            // This is not a cancellation, "spawn" a new server
+            readBuffer.ReadPosition -= 8;
+            var server = new PgServerMock(stream, readBuffer, writeBuffer, ++_processIdCounter);
+            _allServers.Add(server);
+            return new ServerOrCancellationRequest(server);
+        }
+
+        internal async Task<PgServerMock> AcceptServer()
+        {
+            var serverOrCancellationRequest = await Accept();
+            if (serverOrCancellationRequest.Server is null)
+                throw new InvalidOperationException("Expected new server connection but a cancellation request occurred instead");
+            return serverOrCancellationRequest.Server;
         }
 
         internal PgServerMock GetPendingServer()
```

Comment:
Both `GetPendingServer` and `GetPendingCancellationRequest` assume that the server and the cancellation request are already in a queue, which isn't always true. Just something to think about, no need to fix this right now.

Reply:
Yeah. The way I'm currently thinking about it, is that for typical test code a simple NpgsqlConnection.OpenAsync call is awaited - that's where we block - and once that's done the server is supposed to be available. If we run into a scenario that requires something else we can always change this (this mocking infrastructure will definitely evolve).